lib/mongoid/extensions/date.rb

```diff
@@ -37,9 +37,18 @@ module ClassMethods
         #
         # @param [ Time ] object The time from Mongo.
         #
-        # @return [ Date ] The object as a date.
+        # @return [ Date | nil ] The object as a date or nil.
         def demongoize(object)
-          ::Date.new(object.year, object.month, object.day) if object
+          return nil if object.nil?
+          if object.acts_like?(:time) || object.acts_like?(:date)
+            ::Date.new(object.year, object.month, object.day)
+          elsif object.respond_to?(:to_date)
+            begin
+              object.to_date
+            rescue ArgumentError
```

Comment:
Was this theoretical or does something really raise ArgumentError that you know of?

Reply:
Yes:
```
irb(main):001:0> "bogus".to_date
Traceback (most recent call last):
       16: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:25:in `start'
       15: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
       14: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:31:in `dispatch'
       13: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
       12: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
       11: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
       10: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:484:in `exec'
        9: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:23:in `run'
        8: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:58:in `kernel_load'
        7: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:58:in `load'
        6: from /Users/neil.shweky/.gem/bin/irb:23:in `<top (required)>'
        5: from /Users/neil.shweky/.gem/bin/irb:23:in `load'
        4: from /Users/neil.shweky/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        3: from (irb):1
        2: from /Users/neil.shweky/.gem/gems/activesupport-7.0.3/lib/active_support/core_ext/string/conversions.rb:48:in `to_date'
        1: from /Users/neil.shweky/.gem/gems/activesupport-7.0.3/lib/active_support/core_ext/string/conversions.rb:48:in `parse'
Date::Error (invalid date)
```
This raises a `Date::Error` which only dates back to Ruby 2.7, and which is also caught by rescuing an ArgumentError:
```
irb(main):002:1* begin
irb(main):003:1*   "bogus".to_date
irb(main):004:1* rescue ArgumentError
irb(main):005:1*   "rescued"
irb(main):006:0> end
=> "rescued"
```